specgen/specgen.go

```diff
@@ -0,0 +1,518 @@
+// Package specgen provides a utility for generating utls.ClientHelloSpec.
+package specgen
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"reflect"
+
+	tls "github.com/getlantern/utls"
+)
+
+const (
+	// When we print byte slices, we print this many bytes per line.
+	bytesPerLine = 12
+
+	tlsRecordHeaderLen = 5
+)
+
+// WriteHelloSpec writes a utls.ClientHelloSpec corresponding to the input ClientHello.
+//
+// packagePrefix can be used to control the package qualifier on utls definitions. For example, if
+// packagePrefix is provided as "tls", then cipher suites will be specified like
+// "tls.TLS_AES_128_GCM_SHA256" in the output code. If packagePrefix is the empty string, then no
+// qualifier will be used. The example cipher suite would be printed as "TLS_AES_128_GCM_SHA256".
+func WriteHelloSpec(w io.Writer, clientHello []byte, packagePrefix string) error {
+	prefix := ""
+	if packagePrefix != "" {
+		prefix = packagePrefix + "."
+	}
+
+	helloSpec, err := tls.FingerprintClientHello(clientHello[tlsRecordHeaderLen:])
+	if err != nil {
+		return fmt.Errorf("failed to fingerprint hello: %w", err)
+	}
+
+	fmt.Fprintf(w, "%sClientHelloSpec{\n", prefix)
```

Comment:
Am I right this doesn't create a standalone file? I think we want the `DO NOT EDIT` header, a `package whatever`, etc.

Reply:
This does not create a standalone file - just the `ClientHelloSpec` code.  So... I think we don't want that?  I mean, it wouldn't matter a whole lot if there were such a preamble.  But the only real useful part of the output is the `ClientHelloSpec`.  The user is going to just going to remove the preamble.  If this were spit out as a standalone file, what would you name the variable?  You'd be hard-pressed to come up with an intelligent name which is actually useful to the user.  With the current implementation, the user gets only what they came here for.